pipeline/context.go

```diff
@@ -0,0 +1,124 @@
+package pipeline
+
+import (
+	"context"
+	pbsubstreams "github.com/streamingfast/substreams/pb/sf/substreams/v1"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	ttrace "go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+)
+
+type RequestContext struct {
+	ctx     context.Context
+	request *pbsubstreams.Request
+	tracer  ttrace.Tracer
+	logger  *zap.Logger
+
+	traces []*Trace
+}
+
+type Trace struct {
+	name string
+	ctx  context.Context
+	span ttrace.Span
+}
+
+func NewRequestContext(ctx context.Context, req *pbsubstreams.Request, tracer ttrace.Tracer) *RequestContext {
+	return &RequestContext{
+		ctx:     ctx,
+		request: req,
+		tracer:  tracer,
+		logger:  zap.NewNop(),
+	}
+}
+
+func (r *RequestContext) StartSpan(spanName string) {
+	ctx, span := r.tracer.Start(r.ctx, spanName)
+	r.traces = append(r.traces, &Trace{
+		name: spanName,
+		ctx:  ctx,
+		span: span,
+	})
+}
```

Comment:
If we return the `span` here, the caller can `defer reqctx.StartSpan("name").End()`

Reply:
Also, I'd vote for a `currentSpan` and some `panic` if we don't unwind properly, or if we go too far.  This would mean structural errors in our code, and warrants a panic I think.

Reply:
With `currentSpan`, we could still add SetAttributes() to that span, anywhere in the scope of a `StartSpan()` before the `defer ..End()` is called, and it would add it to the current span.